pkg/resource/graph/dependency_graph_rapid_test.go

```diff
@@ -0,0 +1,329 @@
+// Copyright 2016-2021, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Model-checks dependency_graph functionality against simple models
+// using property-based testing.
+//
+// Currently this assumes a simplified model of `resource.State`
+// relevant to dependency calculations; `dependency_graph` only
+// accesses these fields:
+//
+//     type State struct {
+//         Dependencies []resource.URN
+//         URN          resource.URN
+//         Parent       resource.URN
+//         Provider     string
+//         Custom       bool
+//     }
+//
+// At the moment only `Custom=true` (Custom, not Component) resources
+// are tested.
+package graph
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"testing"
+
+	"pgregory.net/rapid"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// Models ------------------------------------------------------------------------------------------
+
+var isParent R = func(child, parent *resource.State) bool {
+	return child.Parent == parent.URN
+}
+
+var hasProvider R = func(res, provider *resource.State) bool {
+	return resource.URN(res.Provider) == provider.URN
+}
+
+var hasDependency R = func(res, dependency *resource.State) bool {
+	for _, dep := range res.Dependencies {
+		if dependency.URN == dep {
+			return true
+		}
+	}
+	return false
+}
+
+var expectedDependenciesOf R = union(isParent, hasProvider, hasDependency)
+
+// Verify `DependneciesOf` against `expectedDependenciesOf`.
+func TestRapidDependenciesOf(t *testing.T) {
+	graphCheck(t, func(t *rapid.T, universe []*resource.State) {
+		dg := NewDependencyGraph(universe)
+		for _, a := range universe {
+			aD := dg.DependenciesOf(a)
+			for _, b := range universe {
+				if isParent(a, b) {
+					assert.Truef(t, aD[b],
+						"DependenciesOf(%v) is missing a parent %v",
+						a.URN, b.URN)
+				}
+				if hasProvider(a, b) {
+					assert.Truef(t, aD[b],
+						"DependenciesOf(%v) is missing a provider %v",
+						a.URN, b.URN)
+				}
+				if hasDependency(a, b) {
+					assert.Truef(t, aD[b],
+						"DependenciesOf(%v) is missing a dependecy %v",
+						a.URN, b.URN)
+				}
+				if aD[b] {
+					assert.True(t, expectedDependenciesOf(a, b),
+						"DependenciesOf(%v) includes an unexpected %v",
+						a.URN, b.URN)
+				}
```

Comment:
I might be confused, but this doesn't seem to provide assertions for transitive dependencies of `a`. It feels like this is asserting `DependenciesOf(x) = list(x.parent, x.Dependencies.flatten(), x.provider).filter(x != nil)`.

Reply:
That's exactly true, if we do not consider component resources ( just custom resources Custom: true). If it provided assertions for transitive dependencies they would fail, because that's not what DependenciesOf computes at the moment.

Reply:
Looking at the DependenciesOf code though, the for loop is awfully confusing, it looks like this was intended to maybe be transitive but isn't (even in the case of ignoring the Component Resource subtleties)....